vending-app/src/main/kotlin/com/google/android/finsky/splitinstallservice/SplitInstallManager.kt

```diff
@@ -0,0 +1,337 @@
+        Log.d(TAG, "requestDownloadUrls languages: $languages")
+        val response = httpClient.get(
+            url = requestUrl.toString(),
+            headers = buildRequestHeaders(authToken, 1, languages).onEach { Log.d(TAG, "key:${it.key}  value:${it.value}") },
+            adapter = GoogleApiResponse.ADAPTER
+        )
+        Log.d(TAG, "requestDownloadUrls end response -> $response")
+        val splitPkgInfoList = response.response?.splitReqResult?.pkgList?.pkgDownLoadInfo ?: throw RuntimeException("splitPkgInfoList is null")
+        val packSet = ArraySet<Triple<String, String, Int>>()
+        splitPkgInfoList.filter {
+            !it.splitPkgName.isNullOrEmpty() && !it.downloadUrl.isNullOrEmpty()
+        }.forEach { info ->
+            packs.filter {
+                it.contains(info.splitPkgName!!)
+            }.forEach {
+                packSet.add(Triple(first = it, second = info.downloadUrl!!, STATUS_DOWNLOADING))
+            }
+        }
+        Log.d(TAG, "requestDownloadUrls end packSet -> $packSet")
+        return packSet.onEach { updateSplitInstallRecord(callingPackage, it) }
+    }
+
+    private suspend fun getOauthToken(): String {
+        val accounts = AccountManager.get(context).getAccountsByType(DEFAULT_ACCOUNT_TYPE)
+        var oauthToken: String? = null
+        if (accounts.isEmpty()) {
+            Log.w(TAG, "No Google account found")
+            throw RuntimeException("No Google account found")
+        } else for (account: Account in accounts) {
+            oauthToken = try {
+                getAuthToken(AccountManager.get(context), account, AUTH_TOKEN_SCOPE).getString(AccountManager.KEY_AUTHTOKEN)
+            } catch (e: AuthenticatorException) {
+                Log.w(TAG, "Could not fetch auth token for account $account")
+                null
+            }
+            if (oauthToken != null) {
+                break
+            }
+        }
+        return oauthToken ?: throw RuntimeException("oauthToken is null")
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun checkSplitInstalled(callingPackage: String, splitName: String): Boolean {
+        if (!splitInstallRecord.containsKey(callingPackage)) return false
+        return splitInstallRecord[callingPackage]?.find { it.first == splitName }?.third != STATUS_UNKNOWN
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun updateSplitInstallRecord(callingPackage: String, triple: Triple<String, String, Int>) {
+        splitInstallRecord[callingPackage]?.let { triples ->
+            val find = triples.find { it.first == triple.first }
+            find?.let { triples.remove(it) }
+            triples.add(triple)
+        } ?: run {
+            val triples = ArraySet<Triple<String, String, Int>>()
+            triples.add(triple)
+            splitInstallRecord[callingPackage] = triples
+        }
+    }
```

Comment:
This method would be entirely unnecessary if `splitInstallRecord` had the data type `Map<String, Map<PackageComponent, DownloadStatus>>` instead of `HashMap<String, ArraySet<Triple<String, String, Int>>>`, with these new classes:

```
data class PackageComponent(
    val componentName: String,
    val url: String
)

enum class DownloadStatus {
    PENDING,
    FAILED,
    COMPLETE
}
```

Reply:
You can find my suggestions here: https://github.com/e-foundation/android_packages_apps_GmsCore/blob/c3398b989ab1a01aaf0bed7a6900cc0f46225ad6/vending-app/src/main/kotlin/com/google/android/finsky/splitinstallservice/SplitInstallManager.kt